src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1905 @@
+  //  haystack and needle are now UTF-16 encoded).
+  bool isReallyUL = isUL;
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         AVX2 code
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  assert(VM_Version::supports_avx2(), "Needs AVX2");
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         Code generation explanation:
+  //
+  //  The generator will generate code for three cases:
+  //  1. Both needle and haystack are Latin-1 (single byte) encoded (LL)
+  //  2. Both the needle and haystack are UTF-16 encoded (two bytes per character) (UU)
+  //  3. The haystack is UTF-16 encoded and the needle is Latin-1 encoded (UL)
+  //
+  //  The case of the haystack being Latin-1 and the needle being UTF-16 is short-circuited
+  //  so that we never get called in this case.
+  //
+  //  For the UL case (haystack UTF-16 and needle Latin-1), the needle will be expanded
+  //  onto the stack (for size <= MAX_NEEDLE_LEN_TO_EXPAND) and the UU code will do the work.
+  //  For UL where the needle size is > MAX_NEEDLE_LEN_TO_EXPAND and the haystack size minus
+  //  the needle size is less than 32 bytes, we default to a
+  //  byte-by-byte comparison (this will be rare).
+  //
+  //  Note that the code assumes MAX_NEEDLE_LEN_TO_EXPAND is >= 32.
+  //
+  //  The UU and LL cases are identical except for the loop increments and loading
+  //  of the characters into registers.  UU loads and compares words, LL - bytes.
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+
+  const Register haystack_p     = c_rarg0;
+  const Register haystack_len_p = c_rarg1;
+  const Register needle_p       = c_rarg2;
+  const Register needle_len_p   = c_rarg3;
+
+  // Addresses of the two jump tables used for small needle processing
+  address big_jump_table;
+  address small_jump_table;
+
+  Label L_begin;
+
+  Label L_returnRBP, L_checkRangeAndReturn, L_returnError, L_bigCaseFixupAndReturn;
+  Label L_bigSwitchTop, L_bigCaseDefault, L_smallCaseDefault;
+  Label L_nextCheck, L_checksPassed, L_zeroCheckFailed, L_return;
+  Label L_wcharBegin, L_continue, L_wideNoExpand;
+
+  __ align(CodeEntryAlignment);
+  fnptrs[ae] = __ pc();
+  __ enter();  // required for proper stackwalking of RuntimeStub frame
+
+  // Check for trivial cases
+  // needle length == 0?
+  __ cmpq(needle_len_p, 0);
+  __ jg_b(L_nextCheck);
+  __ xorq(rax, rax);
```

Comment:
Thanks for finding this.  It was ignorance on my part as I thought the xorq would have logic to not emit the REX prefix if not necessary, but it doesn't.  Fixed.

Reply:
Right, it seems to surprise people. There's a lot of preexisting uses of xorq / xorptr to zero a register. I think it would make sense to implement this logic in xorq. I can do this if others agree.

Reply:
Good idea.  I vote yes.